# Scripts used in lung NKT scRNA-seq project
This site publish R and Python scripts used in our scRNA-seq study.
scRNA-seq was performed using 10x Genomics platgorm and read counts were provided by cellranger.

# Required software
-Cellranger (https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) and reference database provided by the vendor.
-Seurat toolkit for single cell genomics (https://satijalab.org/seurat/)
-Following Python packages.
--Numpy (version >=1.19)
--Pandas (version >= 1.5)
--Plotly (version >=5.0)
--Numba (version >=0.49)
--Reportlab (version >=3)
--scikit-learn (version >=1.1)
--scikit-image (version >=0.17)
--Scipy (version >=1.9)

Python packages are installed with the pip command as `pip install numpy scipy pandas plotly numba reportlab scikit-image scikit-learn scipy`.

You can follow instruction of providers sites to install Cellranger and Seurat.


# Processing scRNA-seq data
Fastq files were generated using bcl2fastq (version 2.20.0.422) and processed using Cellranger (version 6.1.2).
Filtered matrix generated by Cellranger were successfully processed using Seurat (version 4.0.3) with R script seurat.R as follows.

    Rscript seurat.R [filtered_matrix_directory] [output_directory]

This script uses

# Detection of viable cells using MT genes and read counts.
Since each cell has viriable counts of total mRNAs and mitochondrial RNAs and we do not always identify an individual cell as dead or duplet or viable.

    python draw_figures.py cellstate --seurat-dir [seurat.R output directory] -o [output directory] [--verbose]

# Trajectory analysis

    python mstpath.py --seurat-dir [seurat.R output directory] -i [anova_cluster.py output file] --organism ]mm10/hg38] 
